### PR TITLE
Parameterized B-field (+ more TTree validation with MPlex fitting)

### PR DESCRIPTION
**Pull request by kmcdermo**
The main purpose of this PR is to include the parameterized magnetic field in the simulation of the toyMC, as well as in fitting in MPlex (and also Smatrix fitting for the sake of it).

The calculation is performed once per layer per track, at the start of the propagation.  A simple flag is passed from the given propagation routines to use either the parameterized magnetic field or the static 3.8112 T.  The calculation itself is a second order polynomial in z multiplied by a single scale factor in r^2.  

As a reminder, this is the field strength map compared to the static one:
https://kmcdermo.web.cern.ch/kmcdermo/magfield/fieldmap/

I will attach the full benchmarking shortly.  The first results show no change in time to the fitting, where the calculation actually takes place and is timed:
https://kmcdermo.web.cern.ch/kmcdermo/magfield/

Also included in this PR is a number of additions to the TTree validation, useful for Mplex fitting (and therefore for optimization studies).  By using the command line option "fit-val", and compiling with ROOT, we can make some basic residual+pull plots of the fitting at every layer.  Using this to test the magnetic field parameterization in toyMC, things look as expected:
https://kmcdermo.web.cern.ch/kmcdermo/magfield/physics/
(** plot for pt pull had a bug, now fixed in the latest commit --> ignore the title in the plot)

### CODE REVIEW

**Comment by kmcdermo** [comment]
Plots are now updated in linked directory above.  Benchmarking looks fairly standard -- I see the same problems as @mpbl in the endcap building with excessive speedup because of poor single threaded performance.

**Comment by kmcdermo** [comment]
@dan131riley @osschar I have merged the latest PR from @mpbl, and I can confirm everything compiles.

I will run the benchmarking later, although as already pointed out, benchmarking from before was the same.

File: Hit.h [hunk 1]
```diff
@@ -100,6 +100,10 @@ inline float getEta(float r, float z) {
   return -1.0f * std::log( std::tan(getTheta(r,z)/2.) );
 }
 
+inline float getEta(float theta){
+  return -1.0f * std::log( std::tan(theta/2.) );
```
Comment (dan131riley) [inline Hit.h]: should be 2.f, same for line 100 (that ones my fault, I missed it when doing the scan for doubles)

File: mkFit/PropagationMPlex.icc [hunk 2]
```diff
@@ -45,7 +46,7 @@ static inline void helixAtRFromIterative_impl(const Tf& __restrict__ inPar,
       const float ptinv  = 1.f/pt;
       const float pt2inv = ptinv*ptinv;
       //p=0.3Br => r=p/(0.3*B)
-      const float k = inChg(n, 0, 0) * 100.f / (-Config::sol*Config::Bfield);
+      const float k = inChg.ConstAt(n, 0, 0) * 100.f / (-Config::sol*(useParamBfield?Config::BfieldFromZR(inPar.ConstAt(n,2,0),r0):Config::Bfield));
```
Comment (dan131riley) [inline mkFit/PropagationMPlex.icc]: I believe we decided that use of Matriplex At() and ConstAt() is discouraged in new code, there's an overload of operator() for that.

File: mkFit/PropagationMPlex.icc [hunk 3]
```diff
@@ -251,9 +253,9 @@ static inline void helixAtRFromIterativeCCS_impl(const Tf& __restrict__ inPar,
       errorProp(n,4,4) = 1.f;
       errorProp(n,5,5) = 1.f;
 
-      const float k = inChg(n, 0, 0) * 100.f / (-Config::sol*Config::Bfield);
-      const float r = msRad(n, 0, 0);
       float r0 = hipo(inPar(n, 0, 0), inPar(n, 1, 0));
+      const float k = inChg.ConstAt(n, 0, 0) * 100.f / (-Config::sol*(useParamBfield?Config::BfieldFromZR(inPar.ConstAt(n,2,0),r0):Config::Bfield));
```
Comment (dan131riley) [inline mkFit/PropagationMPlex.icc]: see prev comment wrt ConstAt

**Comment by kmcdermo** [comment]
@dan131riley I made the small changes as requested.  I also made some other changes from double to float in Hit.h as well.

**Comment by kmcdermo** [comment]
dump to this PR.  any word?

**Comment by osschar** [comment]
Sorry, didn't realize you were waiting for me :)

**Pull request merged**